### PR TITLE
An attempt at cleaning up the Poseidon2 API

### PR DESCRIPTION
**Pull request by SyxtonPrime**
Not sure if this is better but was simplest alternative I could come up with.

Moved the constants from the Poseidon2 struct into the internal/external layer structs. Seems to be somewhat nicer though it will involve some conditional compilation as the structs need to be defined inside the NEON/AVX2/AVX512 crates so something else needs to be done if compiling on basic (with no vectorization available).

I've migrated the code for Goldilocks and M31 over to give a sense of the benefits/negatives of this new API and checked that there are no meaningful changes to benchmarks.

The biggest issue with this approach is that in order for things to work if no vectorizations are available, we need to conditionally compile a few extra lines of code. I've put these in a new file for now, very happy to change the name or anything else about that file. (Maybe put it in a folder in case we need to do this in some other situations too?)

### CODE REVIEW

File: mersenne-31/src/x86_64_avx2/poseidon2.rs [hunk 1]
```diff
@@ -1,13 +1,26 @@
...
 
-use crate::{
-    exp5, Mersenne31, PackedMersenne31AVX2, Poseidon2ExternalLayerMersenne31,
-    Poseidon2InternalLayerMersenne31, P, P_AVX2,
-};
+use crate::{exp5, Mersenne31, PackedMersenne31AVX2, P, P_AVX2};
+
+#[derive(Debug, Clone, Default)]
+pub struct Poseidon2InternalLayerMersenne31 {
+    pub(crate) internal_constants: Vec<Mersenne31>,
```
Comment (dlubarov) [inline mersenne-31/src/x86_64_avx2/poseidon2.rs:13]: Do we need to keep the input format, seems like it could be forgotten?

Comment (SyxtonPrime) [reply mersenne-31/src/x86_64_avx2/poseidon2.rs:13]: Hmm, I think we need to keep it if we want to be able to run Poseidon2 on lists of field elements, e.g. [F; 16] and not just lists of packed field elements e.g. [PF; 16].

File: mersenne-31/src/poseidon2.rs [hunk 2]
```diff
@@ -173,34 +162,24 @@ mod tests {
...
+    fn poseidon2_mersenne31<const WIDTH: usize>(input: &mut [F; WIDTH])
     where
-        Poseidon2ExternalLayerMersenne31: ExternalLayer<Mersenne31, WIDTH, D>,
+        Poseidon2ExternalLayerMersenne31<WIDTH>: ExternalLayer<Mersenne31, WIDTH, 5>,
         Poseidon2InternalLayerMersenne31: InternalLayer<
             Mersenne31,
             WIDTH,
-            D,
-            InternalState = <Poseidon2ExternalLayerMersenne31 as ExternalLayer<
+            5,
```
Comment (unzvfu) [inline mersenne-31/src/poseidon2.rs:171]: This magic constant should be given a name, like `SBOX_DEGREE` or similar (same goes for all other occurrences of `5` here).

In fact, given that this value is no longer a parameter that can be set by the caller, maybe it shouldn't be specified here at all, but lower down?

File: goldilocks/src/poseidon2.rs [hunk 3]
```diff
@@ -7,14 +7,24 @@
...
 };
 
 use crate::{to_goldilocks_array, Goldilocks};
 
+pub type Poseidon2Goldilocks<const WIDTH: usize> = Poseidon2<
+    Goldilocks,
+    Poseidon2ExternalLayerGoldilocks<WIDTH>,
+    Poseidon2InternalLayerGoldilocks,
+    WIDTH,
+    7,
```
Comment (unzvfu) [inline goldilocks/src/poseidon2.rs:25]: Not strictly part of this PR, but same comment here for `7` as for `5` in the M31 file: It should be given a name like `const SBOX_DEGREE: usize = 7`.

(It is really a property of the field, so specifying it separately is a bit strange, but I guess it's to allow the use of a more expensive s-box than necessary if someone wants?)

**Review by unzvfu — APPROVED** [review]
Looks good overall, definitely an improvement!

Please do address the 2-element array use, though it's up to you whether to clean up the magic number business here or leave it for later.

File: poseidon2/src/matrix.rs [hunk 4]
```diff
@@ -147,9 +147,17 @@ pub fn mds_light_permutation<
     }
 }
 
+pub trait ExternalLayerConstructor<AF, const WIDTH: usize>
+where
+    AF: AbstractField,
+{
+    /// A constructor which internally will convert the supplied
+    /// constants into the appropriate form for the implementation.
+    fn new_from_constants(external_constants: [Vec<[AF::F; WIDTH]>; 2]) -> Self;
```
[the remark this answers is not in the document]
Comment (SyxtonPrime) [reply poseidon2/src/matrix.rs]: Yep, great point. Just done.

File: goldilocks/src/poseidon2.rs (hunk 3 above)
Comment (SyxtonPrime) [reply goldilocks/src/poseidon2.rs:25]: Hmm yeah I might leave that for now but this is something we should come back to. There is (in some cases) a possible advantage to using a more expensive s-box as it changes the number of internal/external rounds you need. But I'm pretty sure it's never optimal.

**Pull request merged**